### ISSUE TITLE
Make `url()` wrapping more lenient in `background-image` -- even though not officially supported

### ISSUE THREAD

**Issue by benc-uk** [cytoscape/cytoscape.js#1512]
I've found an issue using a data URI in the style for a node.

I'm trying to use embedded/inline SVG content. If I base64 encode it, there is no issue, e.g.
`style:{'background-image': 'url(data:image/svg+xml;base64, ...long base64 string of SVG...)'}`
This works OK

However SVG is a format that is better off not being base64 encoded (it actually ends up larger) so when I try this
`style:{'background-image': 'url(data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="50" height="50" viewBox="0 0 50 50"> ....blah rest of svg file here.... </svg>)'}`
This results in cytoscape.js treating it as an external URL and attempts to make HTTP GET for a very strange URL, which obviously fails.

Omitting the `;utf8` part or trying `;charset=utf-8` has no effect.

**Comment by maxkfranz** [comment]
To validate, try your Cytoscape style in an `<img>`.  The `background-image` property should be the same as for CSS.  Cytoscape just loads images into an `<img>` tag.  So if it doesn't load, there's probably a problem with your image.

In this case, I think you forgot to URI encode your data URI: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI

**Comment by benc-uk** [comment]
I tried it in an plain old `<div class='test'>` and `<img class='test'` with a css class defined as follows
~~~~
.test {
    width: 200px; height: 200px;
    background-image: url('data:image/svg+xml,<svg xmlns="http://www.w3.org/2000/svg" height="50" width="50" viewBox="0 0 50 50"><path fill="#3999C6" d="M49.7 25.7c.5-.5.4-1.3 0-1.8l-2.4-2.4L36.5 11c-.5-.5-1.2-.5-1.7 0s-.6 1.3 0 1.8l11.3 11.1c.5.5.5 1.3 0 1.8L34.6 37.2c-.5.5-.5 1.3 0 1.8s1.3.4 1.7 0L47 28.4l.1-.1 2.6-2.6zM.3 25.7c-.5-.5-.4-1.3 0-1.8l2.4-2.4L13.5 11c.5-.5 1.2-.5 1.7 0s.6 1.3 0 1.8L4.1 23.9c-.5.5-.5 1.3 0 1.8l11.3 11.5c.5.5.5 1.3 0 1.8s-1.3.4-1.7 0L2.8 28.5l-.1-.1-2.4-2.7z"/><path fill="#7FBA00" d="M18.2 24.8c0 1.9-1.6 3.3-3.3 3.3s-3.5-1.6-3.5-3.3 1.4-3.3 3.5-3.3c2 0 3.3 1.6 3.3 3.3zM28.3 24.8c0 1.9-1.6 3.3-3.3 3.3s-3.5-1.6-3.5-3.3 1.6-3.3 3.5-3.3 3.3 1.6 3.3 3.3z"/><circle fill="#7FBA00" cx="35.2" cy="24.8" r="3.3"/></svg>');
}
~~~~
And this displays find in Chrome

**Comment by maxkfranz** [comment]
I meant in `<img src="...">` a valid data URI can be used just as in CSS.  I wasn't really clear.

You will need to encode the URI for proper cross browser support.  Embedding a plain SVG without encoded does not make a valid URI.  I wouldn't be surprised if it works without encoding on some browsers in some cases, but you still ought to encode nonetheless.

**Comment by benc-uk** [comment]
I tried encoding the SVG data and it still doesn't work in Cytoscape, it still thinks it is an external URL and attempts to fetch it over HTTP

~~~~
style:{'background-image': "url('data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20height%3D%2250%22%20width%3D%2250%22%20viewBox%3D%220%200%2050%2050%22%3E%3Cpath%20fill%3D%22%233999C6%22%20d%3D%22M49.7%2025.7c.5-.5.4-1.3%200-1.8l-2.4-2.4L36.5%2011c-.5-.5-1.2-.5-1.7%200s-.6%201.3%200%201.8l11.3%2011.1c.5.5.5%201.3%200%201.8L34.6%2037.2c-.5.5-.5%201.3%200%201.8s1.3.4%201.7%200L47%2028.4l.1-.1%202.6-2.6zM.3%2025.7c-.5-.5-.4-1.3%200-1.8l2.4-2.4L13.5%2011c.5-.5%201.2-.5%201.7%200s.6%201.3%200%201.8L4.1%2023.9c-.5.5-.5%201.3%200%201.8l11.3%2011.5c.5.5.5%201.3%200%201.8s-1.3.4-1.7%200L2.8%2028.5l-.1-.1-2.4-2.7z%22%2F%3E%3Cpath%20fill%3D%22%237FBA00%22%20d%3D%22M18.2%2024.8c0%201.9-1.6%203.3-3.3%203.3s-3.5-1.6-3.5-3.3%201.4-3.3%203.5-3.3c2%200%203.3%201.6%203.3%203.3zM28.3%2024.8c0%201.9-1.6%203.3-3.3%203.3s-3.5-1.6-3.5-3.3%201.6-3.3%203.5-3.3%203.3%201.6%203.3%203.3z%22%2F%3E%3Ccircle%20fill%3D%22%237FBA00%22%20cx%3D%2235.2%22%20cy%3D%2224.8%22%20r%3D%223.3%22%2F%3E%3C%2Fsvg%3E')"}
~~~~

Both the encoded and unencoded SVG data string work when I try ` <img src='data:image/svg+xml,{svg_data_string}'`

**Comment by maxkfranz** [comment]
I looked at the documentation (I should have done this from the start), and I was mistaken:  Nowhere is `url()` wrapping mentioned, and it's unsupported.  It's only a coincidence that some `url()` wrapping works, probably because it was under consideration in some very early version.

I've added a patch that makes the check a bit more lenient, but you really ought to use just the URL in the interest of future compatibility.

Without `url()` wrapping, both your SVG and my own test image work fine.
